Sources/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -0,0 +1,456 @@
+      if src.tryEat(sequence: "&&") { return .symmetricDifference }
+      return nil
+    }
+  }
+
+  /// Consume an escaped atom, starting from after the backslash
+  ///
+  ///     Escaped          -> KeyboardModified | Builtin
+  ///                       | UniScalar | Property
+  ///
+  /// TODO: references
+  mutating func expectEscaped(
+    isInCustomCharacterClass ccc: Bool
+  ) throws -> Value<Atom> {
+    try recordLoc { src in
+      // Keyboard control/meta
+      if src.tryEat("c") || src.tryEat(sequence: "C-") {
+        return .keyboardControl(try src.expectASCII().value)
+      }
+      if src.tryEat(sequence: "M-\\C-") {
+        return .keyboardMetaControl(try src.expectASCII().value)
+      }
+      if src.tryEat(sequence: "M-") {
+        return .keyboardMeta(try src.expectASCII().value)
+      }
+
+      let char = src.eat()
+
+      // Single-character builtins
+      if let builtin = Atom.EscapedBuiltin(
+        char, inCustomCharacterClass: ccc
+      ) {
+        return .escaped(builtin)
+      }
+
+      switch char {
+      // Scalars
+      case "u", "x", "U":
+        return try .scalar(
+          src.expectUnicodeScalar(escapedCharacter: char).value)
+
+      // Unicode property checks
+      case "p", "P":
+        fatalError("TODO: properties")
+
+      case "1"..."9", "g", "k":
+        fatalError("TODO: References")
+
+      default: return .char(char)
+      }
+    }
+  }
+
+
+  /// Try to consume an Atom. 
+  ///
+  ///     Atom             -> SpecialCharacter | POSIXSet
+  ///                       | '\' Escaped | [^')' '|']
+  ///     SpecialCharacter -> '.' | '^' | '$'
+  ///     POSIXSet         -> '[[:' name ':]]'
+  ///
+  /// If `SyntaxOptions.nonSemanticWhitespace` is enabled, also accepts:
+  ///
+  ///     ModernGroupStart -> '(_:'
+  ///
+  mutating func lexAtom(
+    isInCustomCharacterClass customCC: Bool
+  ) throws -> Value<Atom>? {
+    try recordLoc { src in
+      // Check for not-an-atom, e.g. parser recursion termination
+      if src.isEmpty || src.peek() == ")" || src.peek() == "|" {
+        return nil
+      }
+      // TODO: Store customCC in the atom, if that's useful
+
+      // POSIX named set
+      if src.tryEat(sequence: "[[:") {
+        // ...
+        try src.expect(sequence: ":]]")
+        fatalError("TODO: posix named sets")
+      }
+
+      let char = src.eat()
+      switch char {
+      case ")", "|": fatalError("unreachable")
+
+      // (sometimes) special metacharacters
+      case ".": return customCC ? .char(".") : .any
+      case "^": return customCC ? .char("^") : .startOfLine
+      case "$": return customCC ? .char("$") : .endOfLine
+
+      // Escaped
+      case "\\": return try src.expectEscaped(
+          isInCustomCharacterClass: customCC).value
+
+      // TODO: backreferences et al here?
+
+      case "]":
+        assert(!customCC, "parser should have prevented this")
+        fallthrough
```

Comment:
PCRE lets you have e.g `]]]` which are treated as literal, is it expected the parser will guard against this, or should we produce `.char` here?

Reply:
But those are outside custom char classes, right? Otherwise the first one terminates it, as PCRE doesn't do nesting.

Reply:
Ah whoops sorry I totally misread the assertion here :)